### DISCUSSION TITLE
Check pushed file changes with git diff-tree in GitHub Actions

### DISCUSSION THREAD

**Discussion by hankehly** [community/community#25950] [Actions]
<p>My goal is to fetch a list of files that were modified between 2 commits (or in 1 commit) using the <code>git diff-tree</code> plumbing command, and I want to do this inside a GitHub Actions workflow on the ubuntu-latest runtime.</p>
<p>The problem is that the <code>git diff-tree</code> stdout / stderr never appears on screen, and I can’t pipe it to a file. I tried running the command in a step run block, python script and inside a private Docker container action to no avail. <a href="https://github.community/t/find-what-files-changed-in-a-pushed-commit/17037">Another post</a> describes a solution, but this did not work for me. What’s interesting is that I can see the output of other git commands like <code>git --version</code> or even <code>git help diff-tree</code> but not the output of <code>git diff-tree ...</code> itself.</p>
<p>My actions workflow configuration is basic:</p>
<ul>
<li>ubuntu-latest runtime</li>
<li>push events only</li>
<li>private repository</li>
</ul>
<p>Is GitHub preventing me from using the <code>git diff-tree</code> command? If so, where is this documented? Has anyone else encountered this issue?</p>
<p>Here is a snippet of the bash script I want to execute.</p>
<pre><code>...

tmpfile=$(mktemp)

start=$(jq --raw-output .commits[0].id "$GITHUB_EVENT_PATH")
end=$(jq --raw-output .commits[-1].id "$GITHUB_EVENT_PATH")

if ["$start" = "$end"]; then
    git diff-tree --no-commit-id --name-only -r "$start" &gt;"$tmpfile"
else
    git diff-tree --no-commit-id --name-only -r "$start" "$end" &gt;"$tmpfile"
fi

...
</code></pre>
<p>Here is an example of what I see in the workflow console.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/47291df70d5cc2edab8d432bd0b726c8e7c685f9.png" data-download-href="/uploads/short-url/a9vZA9N879Tpu5v9NaVJm0ZVbMB.png?dl=1" title="Screen Shot 2019-12-18 at 15.16.47.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/11639738/181094279-a45f1cad-a889-4c89-860d-b7d66156f5ab.png" alt="Screen Shot 2019-12-18 at 15.16.47.png" data-base62-sha1="a9vZA9N879Tpu5v9NaVJm0ZVbMB" width="690" height="321" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/4/47291df70d5cc2edab8d432bd0b726c8e7c685f9_2_690x321.png, https://aws1.discourse-cdn.com/github/original/2X/4/47291df70d5cc2edab8d432bd0b726c8e7c685f9.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/4/47291df70d5cc2edab8d432bd0b726c8e7c685f9.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/47291df70d5cc2edab8d432bd0b726c8e7c685f9_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-12-18 at 15.16.47.png</span><span class="informations">1014×472 88 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by weide-zhou** [comment, marked as answer]
<p>Please remove/reset the ‘fetch-depth’ for ‘actions/checkout’, output works fine then for the command ‘git diff-tree’. </p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/60dedcb2221f3a8fb18a3aab65072d8f46b50e2d.png" data-download-href="/uploads/short-url/dOXjEwTpojhYUIW8TDOfANB3LEF.png?dl=1" title="git diff-tree1.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181094280-04d929ba-ada2-4c2f-b941-6b54d629bb54.png" alt="git diff-tree1.png" data-base62-sha1="dOXjEwTpojhYUIW8TDOfANB3LEF" width="690" height="109" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/60dedcb2221f3a8fb18a3aab65072d8f46b50e2d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">git diff-tree1.png</span><span class="informations">755×120 5.42 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You can get/pipe the changed files via command below:</p>
<pre><code>- name: get changed files
        id: getfile
        run: |
          echo "::set-output name=files::$(git diff-tree --no-commit-id --name-only -r ${{ github.sha }} | xargs)"

      - name: echo output
        run: |
          echo ${{ steps.getfile.outputs.files }}
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/419bfc11f92284c0fc506fd35a6f8b8abd78f809.png" data-download-href="/uploads/short-url/9mpekxfYseA196Yw687OxdDNyRz.png?dl=1" title="git diff-tree2.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181094281-2c3d6d73-8973-41d0-a070-6a8918d2dce1.png" alt="git diff-tree2.png" data-base62-sha1="9mpekxfYseA196Yw687OxdDNyRz" width="690" height="143" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/419bfc11f92284c0fc506fd35a6f8b8abd78f809_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">git diff-tree2.png</span><span class="informations">1076×224 8.12 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by hankehly** [comment]
<p>Thank you for the quick response and making the extra effort of looking inside one of my repositories. A fetch-depth of <code>1</code> is definitely the issue here. Increasing it solved my issue.</p>

**Comment by murciaelx** [comment]
<p>Hi <a class="mention" href="/weide-zhou">@weide-zhou</a>, thanks for your response.</p>
<p>This however does not work for workflow where 3 coommits were added, you will only get the latest information.</p>
<p>Is there a way to get the global changes of the commits you uploaded to the branch?</p>

**Comment by murciaelx** [comment]
<p>Solved with:</p>
<pre><code>git diff --name-only ${{ github.event.before }} ${{ github.sha }}
</code></pre>

**Comment by BlackSwan2019** [comment]
<p>I am trying to compare against the target branch instead of previous commit. Is there something that would work like this? </p>
<pre><code>git diff --name-only ${{ github.base_ref}} ${{ github.sha }}
</code></pre>
<p>I want to always check the diff in absolute terms against the target branch, not just between commits. Thanks.</p>

**Comment by shotor** [comment]
<p>This almost works for me. I’m trying to find occurences of a string. Maybe you can help.</p>
<pre><code>echo "::set-output name=grepValue::$( git diff --name-only ${{ github.event.before }}..${{ github.sha }} | grep -c '^proxy' )"
</code></pre>
<p>This works fine locally but in the builds I get an error (with dots):</p>
<pre><code>fatal: Invalid revision range 9190c0ff82b6b517961997f75090bf9128eea10d..6a5d620bd047d53f38bd5e2966ef63acb1d2dee6
</code></pre>
<p>And without dots:</p>
<pre><code>fatal: bad object 603bf6e0d69a3680413eb48049bb1f9644069965
</code></pre>

**Comment by murciaelx** [comment]
<p>Can you send link to repo/yaml?</p>

**Comment by pozil** [comment]
<p>In case you’re still looking for answer <a class="mention" href="/shotor">@shotor</a> or if someone has the same issue, this “bad object” error is happening because of the default config for <a href="https://github.com/actions/checkout" rel="nofollow noopener">actions/checkout@v2</a>:</p>
<blockquote>
<p>Only a single commit is fetched by default, for the ref/SHA that triggered the workflow. Set  <code>fetch-depth: 0</code>  to fetch all history for all branches and tags.</p>
</blockquote>
<p>This means that you need do something like this:</p>
<pre><code class="lang-auto"># Checkout the source code
- name: 'Checkout source code'
  uses: actions/checkout@v2
  with:
     fetch-depth: 0

# Check for changes
- name: Check for changes
   id: checkForChanges
   run: |
      echo "::set-output name=grepValue::$( git diff --name-only ${{ github.event.before }} ${{ github.sha }} | grep -c '^proxy' )"
</code></pre>

**Comment by ArtemSBulgakov** [comment]
<p>Previous answer (by <a class="mention" href="/pozil">@pozil</a>) fetches full repository and it may be <strong>slow on big repositories</strong>. My solution is to <strong>fetch with depth 1 by using <code>actions/checkout</code> and fetch later manually</strong>. Also my solution works for <strong>both Pull Requests and normal pushes</strong> to branches.</p>
<pre><code class="lang-yaml">on: [push, pull_request]

jobs:
  job1:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Check for changes
        id: diff
        run: |
          if [ $GITHUB_BASE_REF ]; then
            # Pull Request
            git fetch origin $GITHUB_BASE_REF --depth=1
            export DIFF=$( git diff --name-only origin/$GITHUB_BASE_REF $GITHUB_SHA )
            echo "Diff between origin/$GITHUB_BASE_REF and $GITHUB_SHA"
          else
            # Push
            git fetch origin ${{ github.event.before }} --depth=1
            export DIFF=$( git diff --name-only ${{ github.event.before }} $GITHUB_SHA )
            echo "Diff between ${{ github.event.before }} and $GITHUB_SHA"
          fi
          echo "$DIFF"
          # Escape newlines (replace \n with %0A)
          echo "::set-output name=diff::$( echo "$DIFF" | sed ':a;N;$!ba;s/\n/%0A/g' )"
      - run: echo "${{ steps.diff.outputs.diff }}"
</code></pre>
<p>CPython <a href="https://github.com/python/cpython/blob/38d3864efe914fda64553e2ec75c9ec15574483f/.github/workflows/build.yml#L21-L36" rel="nofollow noopener">uses similar way</a> to run tests only if changed files are not in Docs directory (for Pull Requests). For pushes to master they always run tests.</p>
<p><strong>Update:</strong><br>
<a href="https://stackoverflow.com/a/62953566/11948346" rel="nofollow noopener">Answering to the StackOverflow question</a> I wrote a script that detects changed directories and <strong>generates matrix for next job</strong>.</p>
<pre><code class="lang-auto">name: Build
on: [push, pull_request]

jobs:

  generate-matrix:
    name: Generate matrix for build
    runs-on: ubuntu-latest
    outputs:
      matrix: ${{ steps.set-matrix.outputs.matrix }}
    steps:
      - uses: actions/checkout@v2
      - name: Check changed files
        id: diff
        run: |
          # See https://github.community/t/check-pushed-file-changes-with-git-diff-tree-in-github-actions/17220/10
          if [ $GITHUB_BASE_REF ]; then
            # Pull Request
            git fetch origin $GITHUB_BASE_REF --depth=1
            export DIFF=$( git diff --name-only origin/$GITHUB_BASE_REF $GITHUB_SHA )
            echo "Diff between origin/$GITHUB_BASE_REF and $GITHUB_SHA"
          else
            # Push
            git fetch origin ${{ github.event.before }} --depth=1
            export DIFF=$( git diff --name-only ${{ github.event.before }} $GITHUB_SHA )
            echo "Diff between ${{ github.event.before }} and $GITHUB_SHA"
          fi
          echo "$DIFF"
          # Escape newlines (replace \n with %0A)
          echo "::set-output name=diff::$( echo "$DIFF" | sed ':a;N;$!ba;s/\n/%0A/g' )"
      - name: Set matrix for build
        id: set-matrix
        run: |
          # See https://stackoverflow.com/a/62953566/11948346
          DIFF="${{ steps.diff.outputs.diff }}"
          JSON="{\"include\":["

          # Loop by lines
          while read path; do
            # Set $directory to substring before /
            directory="$( echo $path | cut -d'/' -f1 -s )"

            if [ -z "$directory" ]; then
              continue # Exclude root directory
            elif [ "$directory" == docs ]; then
              continue # Exclude docs directory
            elif [ "$path" == *.rst ]; then
              continue # Exclude *.rst files
            fi

            # Set $os. "ubuntu-latest" by default. if directory starts with windows, then "windows-latest"
            os="ubuntu-latest"
            if [ "$directory" == windows* ]; then
              os="windows-latest"
            fi

            # Add build to the matrix only if it is not already included
            JSONline="{\"directory\": \"$directory\", \"os\": \"$os\"},"
            if [[ "$JSON" != *"$JSONline"* ]]; then
              JSON="$JSON$JSONline"
            fi
          done &lt;&lt;&lt; "$DIFF"

          # Remove last "," and add closing brackets
          if [[ $JSON == *, ]]; then
            JSON="${JSON%?}"
          fi
          JSON="$JSON]}"
          echo $JSON

          # Set output
          echo "::set-output name=matrix::$( echo "$JSON" )"

  build:
    name: Build "${{ matrix.directory }}" on ${{ matrix.os }}
    needs: generate-matrix
    strategy:
      matrix: ${{fromJson(needs.generate-matrix.outputs.matrix)}}
    runs-on: ${{ matrix.os }}
    steps:
      - uses: actions/checkout@v2
      - name: Build
        run: |
          cd ${{ matrix.directory }}
          echo "${{ matrix.directory }} ${{ matrix.os }}"
</code></pre>

**Comment by Frank-Reichenbach** [comment]
Due to some issue I had on PR's I want to share my solution for the PR part (PUSH is CP from a solution above).
```shell
if [ $GITHUB_BASE_REF ]; then
    # Pull Request
    # 'github.event.before/after' and 'github.event.pull_request.base.sha' are inconsistent for PRs 
    # Find the commit where the PR branch was forked from the base branch
    git fetch origin ${{ github.base_ref }} ${{ github.event.pull_request.head.ref }}
    FORK_POINT=$( git show-branch --merge-base "origin/${{ github.base_ref }}" "origin/${{ github.event.pull_request.head.ref }}" )
    LATEST_COMMIT_IN_PR=${{ github.event.pull_request.head.sha }}
    CHANGED_FILES=$( git diff --name-only $FORK_POINT $LATEST_COMMIT_IN_PR )
    echo "Changes for PR"
    echo "Diff between $FORK_POINT and $LATEST_COMMIT_IN_PR"
else
    # Push
    # 'github.event.before' and 'after' are including all commits of the current push event
    CHANGED_FILES=$(git diff --name-only ${{ github.event.before }} ${{ github.event.after }})
    echo "Changes for PUSH" 
    echo "Diff between ${{ github.event.before }} and ${{ github.event.after }}"
fi
# Print result for verification
echo Changed files:
echo "$CHANGED_FILES"
```
